README.md
[![Build Status](https://travis-ci.org/elmadev/elma-rust.svg?branch=master)](https://travis-ci.org/elmadev/elma-rust) [![Build status](https://ci.appveyor.com/api/projects/status/0mqnblvhjwlyltkn/branch/master?svg=true)](https://ci.appveyor.com/project/hexjelly/elma-rust/branch/master) [![Coverage Status](https://coveralls.io/repos/github/elmadev/elma-rust/badge.svg?branch=master)](https://coveralls.io/github/elmadev/elma-rust?branch=master) [![Docs](https://docs.rs/elma/badge.svg)](https://docs.rs/elma/)

# ![logo](http://i.imgur.com/4Pg7LyG.png)

[Elasto Mania](http://elmaonline.net/) file handler crate for Rust.
Very much still a work in progress.

## Requirements

Until a 1.0 release this will only target the latest stable version of Rust.

## Installation

Add this in your Cargo.toml file:

```toml
[dependencies]
elma = "*"
```

## Documentation

[https://docs.rs/elma/](https://docs.rs/elma/)

## Progress

-   [ ] Across support
-   [x] Elma support
    -   [x] Level
    -   [x] Replay
    -   [x] LGR
    -   [x] full state.dat support

## Usage examples

### Level operations

To create a new default level:

```rust
extern crate elma;
use elma::lev::*;

fn main () {
    let mut level = Level::new();
    level.save("example.lev", Top10Save::No).unwrap();
}
```

![Screenshot of default level](http://i.imgur.com/TGSo1h4.png)


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## \[0.1.13\] - 2018-08-16

### Added

-   Level's object struct now has `.is_player()` and `is_apple()` methods, thanks to Smibu.

### Breaking

-   All main structs now keep track of file paths in a `path` field.

## \[0.1.12\] - 2018-07-17

### Added

-   Added full state.dat support, once again thanks to Smibu <3.

## \[0.1.11\] - 2018-07-14

### Fixed

-   Replays can now be accurately loaded/saved on a byte level thanks to Smibu <3.

### Added

-   Added `.to_bytes()` for `Replay` struct.

## \[0.1.10\] - 2018-05-15

### Breaking

-   `to_parts()` on `Time` now returns tuple with `bool` signifying positive
or negative number, along with positive integers for all parts.

## \[0.1.9\] - 2018-05-12

### Added

-   LGR reading and writing.
-   Impl Deref to i32 for `Time`.

## \[0.1.8\] - 2018-05-07

### Fixed

-   `Time` default display format changed to `00:00,00`.

## \[0.1.7\] - 2018-05-07

### Fixed

-   `Time::from()` works with `&str`s properly.
-   Removed a rogue print statement.

## \[0.1.6\] - 2018-05-07

### Added

-   Added `.from_bytes()` for `Level`, `Replay` and `State` structs.

## \[0.1.5\] - 2018-05-07

### Fixed

-   No longer necessary to have mutable `Level` in order to use `.save()` method.

### Breaking

-   Removed pub `raw` field from `Level` struct.

## \[0.1.4\] - 2018-05-07

### Added

-   Added preliminary state.dat read/write support for best times.
-   Added `Time` struct with methods for converting string to `i32` represented time, and vice versa.

### Breaking

-   Changed `.get_raw()` to `.to_bytes()` on Level struct.

## \[0.1.3\] - 2016-12-28

### Added

-   Added very basic topology checking.

### Fixed

-   All missing documentation.
-   Rewrote error handling slightly.
-   Refactored cargo structure.

## \[0.1.2\] - 2016-06-23

### Fixed

-   Remove slice pattern feature, and use str for matching instead.

## \[0.1.1\] - 2016-06-22

### Fixed

-   Match against slice to fix error with nightly.

## \[0.1.0\] - 2016-06-22

### Added

-   First release.


tests/lgr.rs
extern crate elma;

use elma::lgr::*;
use std::env;

#[test]
fn correctly_loads_lgr_1() {
    let lgr = LGR::load("tests/assets/lgr/Default.lgr");
    assert!(lgr.is_ok());
}

#[test]
fn correctly_loads_lgr_2() {
    let lgr = LGR::load("tests/assets/lgr/Across.lgr");
    assert!(lgr.is_ok());
}

#[test]
fn correctly_loads_saves_and_reloads_lgrs() {
    let mut dir = env::temp_dir();
    dir.push("resaved.lgr");

    let mut orig_default_lgr = LGR::load("tests/assets/lgr/Default.lgr").unwrap();
    orig_default_lgr.save(&dir).unwrap();
    let reloaded_default_lgr = LGR::load(&dir).unwrap();
    assert_eq!(orig_default_lgr, reloaded_default_lgr);

    let mut orig_across_lgr = LGR::load("tests/assets/lgr/Across.lgr").unwrap();
    orig_across_lgr.save(&dir).unwrap();
    let reloaded_across_lgr = LGR::load(&dir).unwrap();
    assert_eq!(orig_across_lgr, reloaded_across_lgr);
}


src/lgr.rs
use byteorder::{ReadBytesExt, WriteBytesExt, LE};
use std::fs;
use std::path::PathBuf;

use super::{
    utils::{string_null_pad, trim_string}, Clip, ElmaError,
};

// Magic arbitrary number to signify start of LGR file.
const LGR: i32 = 0x00_00_03_EA;
// Magic arbitrary number to signify end of LGR file.
const LGR_EOF: i32 = 0x0B_2E_05_E7;

/// LGR related errors.
#[derive(Debug, PartialEq, Eq, Clone, Ord, PartialOrd)]
pub enum LGRError {
    /// Error parsing version.
    InvalidVersion(Vec<u8>),
    /// Invalid header.
    InvalidHeader(i32),
    /// Invalid clipping.
    InvalidClip(u32),
    /// Invalid transparency.
    InvalidTransparency(u32),
    /// Error parsing PictureType.
    InvalidPictureType(u32),
}

/// LGR structure.
#[derive(Default, Debug, Clone, Eq, PartialEq)]
pub struct LGR {
    /// Path of LGR file.
    pub path: Option<PathBuf>,
    /// List of pictures.
    pub picture_list: Vec<Picture>,
    /// Picture data.
    pub picture_data: Vec<PictureData>,
}

/// LGR picture structure.
#[derive(Default, Debug, Clone, Eq, PartialEq)]
pub struct Picture {
    /// Picture name.
    pub name: String,
    /// Picture type.
    pub picture_type: PictureType,
    /// Default distance, 1-999.
    pub distance: u16,
    /// Default clipping.
    pub clipping: Clip,
    /// Transparency.
    pub transparency: Transparency,
}

/// LGR picture data structure.
#[derive(Default, Debug, Clone, Eq, PartialEq)]
pub struct PictureData {
    /// Picture name.
    pub name: String,
    /// Picture data.
    pub data: Vec<u8>,
}

/// Picture types.
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum PictureType {
    /// Normal picture.
    Normal = 100,
    /// Texture.
    Texture = 101,
    /// Mask for textures.
    Mask = 102,
}

impl Default for PictureType {
    fn default() -> Self {
        PictureType::Normal
    }
}

/// Transparency.
#[derive(Debug, Clone, Copy, Eq, PartialEq)]
pub enum Transparency {
    /// No transparency. Only valid for ´Mask´ picture types.
    Solid = 10,
    /// Palette index 0 is transparent color.
    Palette = 11,
    /// Top left pixel is transparent color.
    TopLeft = 12,
    /// Top right pixel is transparent color.
    TopRight = 13,
    /// Bottom left pixel is transparent color.
    BottomLeft = 14,
    /// Bottom right pixel is transparent color.
    BottomRight = 15,
}

impl Default for Transparency {
    fn default() -> Self {
        Transparency::TopLeft
    }
}

impl LGR {
    /// Creates a new LGR.
    pub fn new() -> Self {
        Self::default()
    }

    /// Loads a LGR from file.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::lgr::*;
    /// let lgr = LGR::load("default.lgr").unwrap();
    /// ```
    pub fn load<P: Into<PathBuf>>(path: P) -> Result<Self, ElmaError> {
        let path = path.into();
        let buffer = fs::read(path.as_path())?;
        let mut lgr = Self::parse_lgr(&buffer)?;
        lgr.path = Some(path);
        Ok(lgr)
    }

    /// Load a LGR from bytes.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::lgr::*;
    /// let buffer = &[0,1,2,3,4]; // pretend this is an actual lgr file.
    /// let lgr = LGR::from_bytes(buffer).unwrap();
    /// ```
    pub fn from_bytes<B: AsRef<[u8]>>(buffer: B) -> Result<Self, ElmaError> {
        Self::parse_lgr(buffer.as_ref())
    }

    fn parse_lgr(buffer: &[u8]) -> Result<Self, ElmaError> {
        let mut lgr = Self::new();

        let (version, mut buffer) = buffer.split_at(5);
        // there are no other LGR versions possible, so no need to store it (?)
        if version != b"LGR12" {
            return Err(ElmaError::InvalidLGRFile(LGRError::InvalidVersion(
                version.to_vec(),
            )));
        }

        let picture_len = buffer.read_u32::<LE>()? as usize;
        let expected_header = buffer.read_i32::<LE>()?;
        if expected_header != LGR {
            return Err(ElmaError::InvalidLGRFile(LGRError::InvalidHeader(
                expected_header,
            )));
        }

        // picture.lst section
        let list_len = buffer.read_u32::<LE>()? as usize;
        lgr.parse_list_data(&buffer, list_len)?;
        let (_, buffer) = buffer.split_at(26 * list_len);

        // pcx data
        let bytes_read = lgr.parse_picture_data(&buffer, picture_len)?;

        let (_, mut expected_eof) = buffer.split_at(bytes_read);

        let expected_eof = expected_eof.read_i32::<LE>()?;
        if expected_eof != LGR_EOF {
            return Err(ElmaError::EOFMismatch);
        }

        Ok(lgr)
    }

    fn parse_list_data(&mut self, buffer: &[u8], len: usize) -> Result<(), ElmaError> {
        let (names, buffer) = buffer.split_at(len * 10);
        let (mut picture_types, buffer) = buffer.split_at(len * 4);
        let (mut distances, buffer) = buffer.split_at(len * 4);
        let (mut clippings, buffer) = buffer.split_at(len * 4);
        let (mut transparencies, _) = buffer.split_at(len * 4);

        for n in 0..len {
            let name = trim_string(&names[10 * n..(10 * n) + 10])?;
            let picture_type = match picture_types.read_u32::<LE>()? {
                100 => PictureType::Normal,
                101 => PictureType::Texture,
                102 => PictureType::Mask,
                e => return Err(ElmaError::InvalidLGRFile(LGRError::InvalidPictureType(e))),
            };
            let distance = distances.read_u32::<LE>()? as u16;
            let clipping = match clippings.read_u32::<LE>()? {
                0 => Clip::Unclipped,
                1 => Clip::Ground,
                2 => Clip::Sky,
                e => return Err(ElmaError::InvalidLGRFile(LGRError::InvalidClip(e))),
            };
            let transparency = match transparencies.read_u32::<LE>()? {
                10 => Transparency::Solid,
                11 => Transparency::Palette,
                12 => Transparency::TopLeft,
                13 => Transparency::TopRight,
                14 => Transparency::BottomLeft,
                15 => Transparency::BottomRight,
                e => return Err(ElmaError::InvalidLGRFile(LGRError::InvalidTransparency(e))),
            };

            self.picture_list.push(Picture {
                name,
                picture_type,
                distance,
                clipping,
                transparency,
            });
        }
        Ok(())
    }

    fn parse_picture_data(&mut self, mut buffer: &[u8], len: usize) -> Result<usize, ElmaError> {
        let mut bytes_read = 0;
        // pcx data
        for _ in 0..len {
            let (name, remaining) = buffer.split_at(12);
            let name = trim_string(&name)?;
            let (_, remaining) = remaining.split_at(8);
            let (mut bytes_len, remaining) = remaining.split_at(4);
            let bytes_len = bytes_len.read_i32::<LE>()? as usize;
            let data = remaining[..bytes_len].to_vec();

            self.picture_data.push(PictureData { name, data });
            buffer = &buffer[24 + bytes_len..];
            bytes_read += 24 + bytes_len;
        }
        Ok(bytes_read)
    }

    /// Returns a Vec with bytes representing the LGR as a buffer.
    ///
    /// # Examples
    ///
    /// ```rust
    /// # use elma::lgr::*;
    /// let lgr = LGR::new();
    /// let buffer = lgr.to_bytes().unwrap();
    /// ```
    pub fn to_bytes(&self) -> Result<Vec<u8>, ElmaError> {
        let mut bytes = vec![];
        bytes.extend_from_slice(b"LGR12");
        bytes.write_u32::<LE>(self.picture_data.len() as u32)?;
        bytes.write_i32::<LE>(LGR)?;
        bytes.extend_from_slice(&self.write_picture_list()?);
        bytes.extend_from_slice(&self.write_picture_data()?);
        bytes.write_i32::<LE>(LGR_EOF)?;

        Ok(bytes)
    }

    fn write_picture_list(&self) -> Result<Vec<u8>, ElmaError> {
        let mut bytes = vec![];
        bytes.write_u32::<LE>(self.picture_list.len() as u32)?;
        let mut names = vec![];
        let mut picture_types = vec![];
        let mut distances = vec![];
        let mut clippings = vec![];
        let mut transparencies = vec![];

        for picture in &self.picture_list {
            names.extend_from_slice(&string_null_pad(&picture.name, 10)?);
            picture_types.write_u32::<LE>(picture.picture_type as u32)?;
            distances.write_u32::<LE>(u32::from(picture.distance))?;
            clippings.write_u32::<LE>(picture.clipping as u32)?;
            transparencies.write_u32::<LE>(picture.transparency as u32)?;
        }

        bytes.extend_from_slice(&names);
        bytes.extend_from_slice(&picture_types);
        bytes.extend_from_slice(&distances);
        bytes.extend_from_slice(&clippings);
        bytes.extend_from_slice(&transparencies);

        Ok(bytes)
    }

    fn write_picture_data(&self) -> Result<Vec<u8>, ElmaError> {
        let mut bytes = vec![];

        for picture in &self.picture_data {
            bytes.extend_from_slice(&string_null_pad(&picture.name, 20)?);
            bytes.write_u32::<LE>(picture.data.len() as u32)?;
            bytes.extend_from_slice(&picture.data);
        }

        Ok(bytes)
    }

    /// Save the LGR to a file.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::lgr::*;
    /// let mut lgr = LGR::new();
    /// lgr.save("cool.lgr");
    /// ```
    pub fn save<P: Into<PathBuf>>(&mut self, path: P) -> Result<(), ElmaError> {
        let bytes = self.to_bytes()?;
        let path = path.into();
        fs::write(path.as_path(), &bytes)?;
        self.path = Some(path);
        Ok(())
    }
}


src/lib.rs
#![deny(missing_docs)]

//! Library for reading and writing Elasto Mania files.

extern crate byteorder;
extern crate rand;
#[macro_use]
extern crate nom;
#[macro_use]
extern crate itertools;

use std::{io, string};

/// Various constant values used throughout the game and library.
pub mod constants;
/// Read and write Elasto Mania level files.
pub mod lev;
/// Read and write Elasto Mania LGR files.
pub mod lgr;
/// Read and write Elasto Mania replay files.
pub mod rec;
/// Read and write Elasto Mania state.dat files.
pub mod state;
/// Various utility functions.
pub mod utils;

mod shared;
use lgr::LGRError;
pub use shared::{BestTimes, Clip, Position, Time, TimeEntry, Version};

/// General errors.
#[derive(Debug, PartialEq)]
pub enum ElmaError {
    /// Across files are not supported.
    AcrossUnsupported,
    /// Invalid level file.
    InvalidLevelFile,
    /// Invalid level file name.
    InvalidLevelFilename,
    /// Invalid replay file.
    InvalidReplayFile,
    /// Invalid state file.
    InvalidStateFile,
    /// Invalid LGR file.
    InvalidLGRFile(LGRError),
    /// Invalid gravity value.
    InvalidGravity(i32),
    /// Invalid object value.
    InvalidObject(i32),
    /// Invalid clipping value.
    InvalidClipping(i32),
    /// End-of-data marker mismatch.
    EODMismatch,
    /// End-of-file marker mismatch.
    EOFMismatch,
    /// Invalid event value.
    InvalidEvent(u8),
    /// End-of-replay marker mismatch.
    EORMismatch,
    /// Invalid time format.
    InvalidTimeFormat,
    /// Too short padding.
    PaddingTooShort(isize),
    /// String contains non-ASCII characters.
    NonASCII,
    /// Input/output errors from std::io use.
    Io(std::io::ErrorKind),
    /// String errors from std::String.
    StringFromUtf8(usize),
}

impl From<io::Error> for ElmaError {
    fn from(err: io::Error) -> ElmaError {
        ElmaError::Io(err.kind())
    }
}

impl From<string::FromUtf8Error> for ElmaError {
    fn from(err: string::FromUtf8Error) -> ElmaError {
        ElmaError::StringFromUtf8(err.utf8_error().valid_up_to())
    }
}


src/state.rs
use super::{BestTimes, ElmaError};
use byteorder::{WriteBytesExt, LE};
use constants::TOP10_SIZE;
use nom::{le_i32, le_u32, le_u8};
use shared::TimeEntry;
use std::fs;
use std::path::PathBuf;
use std::str;
use utils::boolean;
use utils::null_padded_string;
use utils::to_bool;
use utils::{string_null_pad, write_top10};

const PLAYER_STRUCT_SIZE: usize = 116;
const PLAYERENTRY_PADDING: usize = 38;
const NUM_INTERNALS: usize = 54;
const PLAYER_NAME_SIZE: usize = 15;
const PLAYERENTRY_NAME_SIZE: usize = 16;
const LEVEL_NAME_SIZE: usize = 20;
const NUM_PLAYERS: usize = 50;
const NUM_LEVELS: usize = 90;
const STATE_START: u32 = 200;
const STATE_END: u32 = 123_432_221;
const STATE_END_ALT: u32 = 123_432_112;
const TOP10_ENTRIES: usize = 10;

#[derive(Debug, Clone, PartialEq, Copy)]
/// Play mode.
pub enum PlayMode {
    /// Single player.
    Single = 1,
    /// Multi player.
    Multi = 0,
}

impl Default for PlayMode {
    fn default() -> Self {
        PlayMode::Single
    }
}

#[derive(Debug, Clone, PartialEq, Copy)]
/// Sound optimization.
pub enum SoundOptimization {
    /// Compatibility mode.
    Compatibility = 1,
    /// Best quality mode.
    BestQuality = 0,
}

impl Default for SoundOptimization {
    fn default() -> Self {
        SoundOptimization::BestQuality
    }
}

#[derive(Debug, Clone, PartialEq, Copy)]
/// Video detail.
pub enum VideoDetail {
    /// Low details.
    Low = 0,
    /// High details.
    High = 1,
}

impl Default for VideoDetail {
    fn default() -> Self {
        VideoDetail::High
    }
}

#[derive(Default, Debug, Clone, PartialEq)]
/// Player entry in state.dat.
pub struct PlayerEntry {
    /// Player name.
    pub name: String,
    /// Skipped internals.
    pub skipped_internals: Vec<bool>,
    /// The index of last internal the player has reached so far.
    pub last_internal: i32,
    /// The last played (selected) internal.
    pub selected_internal: i32,
}

#[derive(Default, Debug, Clone, PartialEq)]
/// Key settings of a player.
pub struct PlayerKeys {
    /// Throttle key.
    pub throttle: u32,
    /// Brake key.
    pub brake: u32,
    /// Rotate right key.
    pub rotate_right: u32,
    /// Rotate left key.
    pub rotate_left: u32,
    /// Change direction key.
    pub change_direction: u32,
    /// Toggle navigator key.
    pub toggle_navigator: u32,
    /// Toggle timer key.
    pub toggle_timer: u32,
    /// Toggle show/hide key.
    pub toggle_show_hide: u32,
}

/// State.dat struct
#[derive(Default, Debug, Clone, PartialEq)]
pub struct State {
    /// Path to State file.
    pub path: Option<PathBuf>,
    /// State file version; the only supported value is 200.
    pub version: u32,
    /// Best times lists. state.dat has a fixed-size array of 90 of these.
    pub times: Vec<BestTimes>,
    /// List of players. state.dat has a fixed-size array of 50 of these.
    pub players: Vec<PlayerEntry>,
    /// Name of player A, maximum 14 characters.
    pub player_a_name: String,
    /// Name of player B, maximum 14 characters.
    pub player_b_name: String,
    /// Keys for player A.
    pub player_a_keys: PlayerKeys,
    /// Keys for player B.
    pub player_b_keys: PlayerKeys,
    /// Whether sound is enabled.
    pub sound_enabled: bool,
    /// Sound optimization.
    pub sound_optimization: SoundOptimization,
    /// Play mode.
    pub play_mode: PlayMode,
    /// Whether flag tag mode is enabled.
    pub flagtag: bool,
    /// Whether bikes are swapped.
    pub swap_bikes: bool,
    /// Video detail.
    pub video_detail: VideoDetail,
    /// Whether objects are animated.
    pub animated_objects: bool,
    /// Whether menus are animated.
    pub animated_menus: bool,
    /// Key for increasing screen size.
    pub inc_screen_size_key: u32,
    /// Key for decreasing screen size.
    pub dec_screen_size_key: u32,
    /// Key for taking a screenshot.
    pub screenshot_key: u32,
    /// Name of last edited level.
    pub last_edited_lev_name: String,
    /// Name of last played external level.
    pub last_played_external: String,
}

#[cfg_attr(rustfmt, rustfmt_skip)]
named!(playerkeys<PlayerKeys>,
  do_parse!(
    throttle: le_u32 >>
    brake: le_u32 >>
    rotate_right: le_u32 >>
    rotate_left: le_u32 >>
    change_direction: le_u32 >>
    toggle_navigator: le_u32 >>
    toggle_timer: le_u32 >>
    toggle_show_hide: le_u32 >>
    (PlayerKeys {
      throttle,
      brake,
      rotate_right,
      rotate_left,
      change_direction,
      toggle_navigator,
      toggle_timer,
      toggle_show_hide,
    })
  )
);

#[cfg_attr(rustfmt, rustfmt_skip)]
named!(playerentry<PlayerEntry>,
  do_parse!(
    name: apply!(null_padded_string, PLAYERENTRY_NAME_SIZE) >>
    cond_reduce!(!name.is_empty(), take!(0)) >>
    skipped_internals: many_m_n!(NUM_INTERNALS, NUM_INTERNALS, map!(le_u8, |x| to_bool(i32::from(x)))) >>
    take!(PLAYERENTRY_PADDING) >>
    last_internal: le_i32 >>
    selected_internal: le_i32 >>
    (PlayerEntry {
      name: name.to_string(),
      skipped_internals,
      last_internal,
      selected_internal,
    })
  )
);

#[cfg_attr(rustfmt, rustfmt_skip)]
named_args!(playernames(num: usize)<Vec<&str>>,
  do_parse!(
    names: many_m_n!(num, num, apply!(null_padded_string, PLAYER_NAME_SIZE)) >>
    cond_reduce!(num <= TOP10_ENTRIES, count!(apply!(null_padded_string, PLAYER_NAME_SIZE), TOP10_ENTRIES - num)) >>
    (names)
  )
);

#[cfg_attr(rustfmt, rustfmt_skip)]
named!(top10part<Vec<TimeEntry>>,
  do_parse!(
    num_times: map!(le_u32, |x| x as usize) >>
    times: many_m_n!(num_times, num_times, map!(le_i32, |x| x.into())) >>
    cond_reduce!(num_times <= TOP10_ENTRIES, count!(le_i32, TOP10_ENTRIES - num_times)) >>
    player_a_names: apply!(playernames, num_times) >>
    player_b_names: apply!(playernames, num_times) >>
    (izip!(times, player_a_names, player_b_names).map(|(time, a, b)| TimeEntry {names: (a.to_string(),b.to_string()), time}).collect())
  )
);

#[cfg_attr(rustfmt, rustfmt_skip)]
named!(top10<BestTimes>,
  do_parse!(
    single: top10part >>
    multi: top10part >>
    (BestTimes {
      single,
      multi,
    })
  )
);

#[cfg_attr(rustfmt, rustfmt_skip)]
named!(parse_state<State>,
  do_parse!(
    version: verify!(le_u32, |x| x == STATE_START) >>
    times: many_m_n!(NUM_LEVELS, NUM_LEVELS, top10) >>
    // Using 0 here instead of 1 generates a warning. Workaround with map_opt + opt.
    players: map_opt!(opt!(many_m_n!(1, NUM_PLAYERS, playerentry)), |x: Option<_>| x.or_else(|| Some(vec![]))) >>
    cond_reduce!(players.len() <= NUM_PLAYERS, take!((NUM_PLAYERS - players.len()) * PLAYER_STRUCT_SIZE)) >>
    num_players: le_u32 >>
    cond_reduce!(players.len() == num_players as usize, take!(0)) >>
    player_a_name: apply!(null_padded_string, PLAYER_NAME_SIZE) >>
    player_b_name: apply!(null_padded_string, PLAYER_NAME_SIZE) >>
    sound_enabled: boolean >>
    sound_optimization: map!(le_i32, |i| if i == 0 {SoundOptimization::BestQuality} else {SoundOptimization::Compatibility}) >>
    play_mode: map!(le_i32, |i| if i == 0 {PlayMode::Multi} else {PlayMode::Single}) >>
    flagtag: boolean >>
    swap_bikes: map!(boolean, |x| !x) >>
    video_detail: map!(le_i32, |i| if i == 0 {VideoDetail::Low} else {VideoDetail::High}) >>
    animated_objects: boolean >>
    animated_menus: boolean >>
    player_a_keys: playerkeys >>
    player_b_keys: playerkeys >>
    inc_screen_size_key: le_u32 >>
    dec_screen_size_key: le_u32 >>
    screenshot_key: le_u32 >>
    last_edited_lev_name: apply!(null_padded_string, LEVEL_NAME_SIZE) >>
    last_played_external: apply!(null_padded_string, LEVEL_NAME_SIZE) >>
    verify!(le_u32, |x| x == STATE_END || x == STATE_END_ALT) >>
  (State {
    path: None,
    version,
    times,
    players,
    player_a_name: player_a_name.to_string(),
    player_b_name: player_b_name.to_string(),
    player_a_keys,
    player_b_keys,
    sound_enabled,
    sound_optimization,
    play_mode,
    flagtag,
    swap_bikes,
    video_detail,
    animated_objects,
    animated_menus,
    dec_screen_size_key,
    inc_screen_size_key,
    last_edited_lev_name: last_edited_lev_name.to_string(),
    last_played_external: last_played_external.to_string(),
    screenshot_key,
  })
  )
);

impl State {
    /// Create a new state.dat.
    pub fn new() -> Self {
        State {
            path: None,
            version: 200,
            times: vec![
                BestTimes {
                    single: vec![],
                    multi: vec![],
                };
                90
            ],
            players: vec![],
            player_a_name: "".to_string(),
            player_b_name: "".to_string(),
            player_a_keys: PlayerKeys {
                throttle: 200,
                brake: 208,
                rotate_right: 205,
                rotate_left: 203,
                change_direction: 57,
                toggle_navigator: 47,
                toggle_timer: 20,
                toggle_show_hide: 2,
            },
            player_b_keys: PlayerKeys {
                throttle: 76,
                brake: 80,
                rotate_right: 81,
                rotate_left: 79,
                change_direction: 82,
                toggle_navigator: 48,
                toggle_timer: 21,
                toggle_show_hide: 3,
            },
            sound_enabled: true,
            sound_optimization: SoundOptimization::BestQuality,
            play_mode: PlayMode::Single,
            flagtag: false,
            swap_bikes: false,
            video_detail: VideoDetail::High,
            animated_objects: true,
            animated_menus: true,
            inc_screen_size_key: 13,
            dec_screen_size_key: 12,
            screenshot_key: 23,
            last_edited_lev_name: "".to_string(),
            last_played_external: "".to_string(),
        }
    }

    /// Load a state.dat file.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::state::*;
    /// let state = State::load("state.dat").unwrap();
    /// ```
    pub fn load<P: Into<PathBuf>>(path: P) -> Result<Self, ElmaError> {
        let path = path.into();
        let buffer = fs::read(path.as_path())?;
        let mut state = State::parse(&buffer)?;
        state.path = Some(path);
        Ok(state)
    }

    /// Load a state.dat file from bytes.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::state::*;
    /// let state = State::from_bytes(&[0,1,2]).unwrap();
    /// ```
    pub fn from_bytes<B: AsRef<[u8]>>(buffer: B) -> Result<Self, ElmaError> {
        State::parse(buffer.as_ref())
    }

    fn parse(buffer: &[u8]) -> Result<Self, ElmaError> {
        let mut v = buffer.to_vec();
        {
            let mut buf = &mut v[..];
            crypt_whole_state(&mut buf);
        }
        match parse_state(&v) {
            Ok((_, state)) => Ok(state),
            Err(_) => Err(ElmaError::InvalidStateFile),
        }
    }

    /// Returns state.dat as a stream of bytes.
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::state::*;
    /// let mut state = State::load("state.dat").unwrap();
    /// let buffer = state.to_bytes().unwrap();
    /// ```
    pub fn to_bytes(&self) -> Result<Vec<u8>, ElmaError> {
        let mut buffer = vec![];
        buffer.write_u32::<LE>(STATE_START)?;

        for level in &self.times {
            let top10_bytes = write_top10(&level)?;
            buffer.extend(top10_bytes);
        }
        buffer.extend(vec![0; (NUM_LEVELS - self.times.len()) * TOP10_SIZE]);

        for p in &self.players {
            buffer.extend(string_null_pad(&p.name, PLAYERENTRY_NAME_SIZE)?);
            p.skipped_internals
                .iter()
                .for_each(|&x| buffer.push(x as u8));
            buffer.extend(vec![0; NUM_INTERNALS - p.skipped_internals.len()]);
            buffer.extend_from_slice(&[0; PLAYERENTRY_PADDING]);
            buffer.write_i32::<LE>(p.last_internal)?;
            buffer.write_i32::<LE>(p.selected_internal)?;
        }
        buffer.extend(vec![
            0;
            (NUM_PLAYERS - self.players.len()) * PLAYER_STRUCT_SIZE
        ]);
        buffer.write_u32::<LE>(self.players.len() as u32)?;
        buffer.extend(string_null_pad(&self.player_a_name, PLAYER_NAME_SIZE)?);
        buffer.extend(string_null_pad(&self.player_b_name, PLAYER_NAME_SIZE)?);
        buffer.write_i32::<LE>(self.sound_enabled as i32)?;
        buffer.write_i32::<LE>(self.sound_optimization as i32)?;
        buffer.write_i32::<LE>(self.play_mode as i32)?;
        buffer.write_i32::<LE>(self.flagtag as i32)?;
        buffer.write_i32::<LE>(!self.swap_bikes as i32)?;
        buffer.write_i32::<LE>(self.video_detail as i32)?;
        buffer.write_i32::<LE>(self.animated_objects as i32)?;
        buffer.write_i32::<LE>(self.animated_menus as i32)?;
        for k in &[&self.player_a_keys, &self.player_b_keys] {
            buffer.write_u32::<LE>(k.throttle)?;
            buffer.write_u32::<LE>(k.brake)?;
            buffer.write_u32::<LE>(k.rotate_right)?;
            buffer.write_u32::<LE>(k.rotate_left)?;
            buffer.write_u32::<LE>(k.change_direction)?;
            buffer.write_u32::<LE>(k.toggle_navigator)?;
            buffer.write_u32::<LE>(k.toggle_timer)?;
            buffer.write_u32::<LE>(k.toggle_show_hide)?;
        }
        buffer.write_u32::<LE>(self.inc_screen_size_key)?;
        buffer.write_u32::<LE>(self.dec_screen_size_key)?;
        buffer.write_u32::<LE>(self.screenshot_key)?;
        buffer.extend(string_null_pad(
            &self.last_edited_lev_name,
            LEVEL_NAME_SIZE,
        )?);
        buffer.extend(string_null_pad(
            &self.last_played_external,
            LEVEL_NAME_SIZE,
        )?);
        buffer.write_u32::<LE>(STATE_END)?;
        crypt_whole_state(&mut buffer[..]);
        Ok(buffer)
    }

    /// Save state.dat
    ///
    /// # Examples
    ///
    /// ```rust,no_run
    /// # use elma::state::*;
    /// let mut state = State::load("state.dat").unwrap();
    /// state.save("newstate.dat").unwrap();
    /// ```
    pub fn save<P: Into<PathBuf>>(&mut self, path: P) -> Result<(), ElmaError> {
        let buffer = self.to_bytes()?;
        let path = path.into();
        fs::write(path.as_path(), &buffer)?;
        self.path = Some(path);
        Ok(())
    }
}

fn crypt_whole_state(buf: &mut [u8]) {
    let state_pieces = [
        4,
        61920,
        5800,
        4,
        PLAYER_NAME_SIZE,
        PLAYER_NAME_SIZE,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        4,
        32,
        32,
        4,
        4,
        4,
        LEVEL_NAME_SIZE,
        LEVEL_NAME_SIZE,
    ];
    let mut curr = 0;
    for p in &state_pieces {
        crypt_state(&mut buf[curr..curr + p]);
        curr += p;
    }
}

fn crypt_state(buffer: &mut [u8]) {
    let mut ebp8: i16 = 0x17;
    let mut ebp10: i16 = 0x2636;

    for mut t in buffer.iter_mut() {
        *t ^= (ebp8 & 0xFF) as u8;
        ebp10 = ebp10.wrapping_add((ebp8.wrapping_rem(0xD3F)).wrapping_mul(0xD3F));
        ebp8 = ebp10.wrapping_mul(0x1F).wrapping_add(0xD3F);
    }
}
